crates/egui/src/atomics/atomic_kind.rs

```diff
@@ -0,0 +1,105 @@
+use crate::{Id, Image, SizedAtomicKind, TextStyle, Ui, WidgetText};
+use emath::Vec2;
+use epaint::text::TextWrapMode;
+use std::fmt::Formatter;
+
+/// The different kinds of [`Atomic`]s.
+#[derive(Clone, Default)]
+pub enum AtomicKind<'a> {
+    /// Empty, that can be used with [`AtomicExt::a_grow`] to reserve space.
+    #[default]
+    Empty,
+    Text(WidgetText),
+    Image(Image<'a>),
+
+    /// For custom rendering.
+    ///
+    /// You can get the [`Rect`] with the [`Id`] from [`AtomicLayoutResponse`] and use a
+    /// [`crate::Painter`] or [`Ui::put`] to add/draw some custom content.
+    ///
+    /// Example:
+    /// ```
+    /// # use egui::{AtomicKind, Button, Id, __run_test_ui};
+    /// # use emath::Vec2;
+    /// # __run_test_ui(|ui| {
+    /// let id = Id::new("my_button");
+    /// let response = Button::new(("Hi!", AtomicKind::Custom(id, Vec2::splat(18.0)))).atomic_ui(ui);
+    ///
+    /// let rect = response.custom_rects.get(&id);
+    /// if let Some(rect) = rect {
+    ///     ui.put(*rect, Button::new("⏵"));
+    /// }
+    /// # });
+    /// ```
+    Custom(Id, Vec2),
+}
+
+impl std::fmt::Debug for AtomicKind<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            AtomicKind::Empty => write!(f, "AtomicKind::Empty"),
+            AtomicKind::Text(text) => write!(f, "AtomicKind::Text({})", text.text()),
+            AtomicKind::Image(image) => write!(f, "AtomicKind::Image({image:?})"),
+            AtomicKind::Custom(id, size) => write!(f, "AtomicKind::Custom({id:?}, {size:?})"),
+        }
+    }
+}
```

Comment:
If we `impl Debug for WidgetText`, then maybe we could `#[derive(Debug)]` for `AtomicKind`?